tests/test_model_derivatives.py

```diff
@@ -163,19 +163,27 @@ def f(value):
         h3 = model.H3.value
         stepgen = numdifftools.MaxStepGenerator(abs(h3) / 2)
     elif param == "FB0":
-        stepgen = numdifftools.MaxStepGenerator(np.abs(model.FB0.value) * 1e-2)
```

Comment:
Why aren't FB parameters `float64`s to begin with?

Reply:
I don't think even the rotational frequency needs more precision.

Reply:
Agree.  Maybe that is more general bookkeeping that could be done elsewhere (reduce precision where not needed).  FB0 is set to `longdouble=True` on creation.  I could change that here, or we can just deal with that later.

Reply:
Rotational frequency is one of the only timing parameters that *does* need to be long double (40 years of timing a MSP with 100ns timing precision gives a nominal absolute frequency precision of 100ns/(40yrs) ~ 1e-16, and there are 3 more sig figs before the decimal place!).  However, the FB terms can have numerical stability issues (since we are doing polynomial calculations with terms that can have amplitudes of like 1e-100) and so I think that is why we went with long double with those.  Numerical derivatives make the numerical stability issues even worse.  So I think we should be very careful with this...

Reply:
Careful in the sense that we shouldn't change the precision of the parameter, or even the precision of the derivative test?  If the former then we can just proceed.  If the latter then we need to understand how to do the tests better.

Reply:
The Nyquist frequency for 40 years is about 0.3 nHz. If we have a 700 Hz pulsar, shouldn't the space needed to represent F0 be log2(700 Hz / 0.3 nHz) ~ 41 bits?

Reply:
Ok... If we consider the numerical stability of Horner method, it makes sense to me that we'll need long double precision.

Reply:
Oops.  I forgot a factor of F0.  For 100ns timing precision and a 40-year baseline, if you have a 1.5ms pulsar, then the nominal absolute freq error is roughly deltaPhi / deltaT ~ (100ns/1.5ms) / 40 yrs ~ 5e-14.  So that is 13 or 14 sig figs.  Then there are 3 more sig figs for the freq before the decimal point.  Double precision floats can give you ~15 sig figs (or digits of precision).  And we are needing that precision right now -- if you look at the recent data releases from all the PTAs, the F0 errors for the best pulsars are better than 1e-13, meaning 16 full digits of precision.

As for your question about bits, the number of bits needed in the mantissa must therefore be more than log2(1e16) = 53.15, and IEEE doubles have 52 bits in the mantissa.

(I think it is really important that pulsar students learn stuff like this -- we need to keep teaching them.  :-)

Reply:
Ok... I was thinking the precision was related to the total number of bits, not the bits allotted for the mantissa. Thanks for the explanation.